### ISSUE TITLE
XGBoostClassifier multiclass objective

### ISSUE THREAD

**Issue by paulbir** [getml/getml-community#9]
Currently "objective" parameter for the `XGBoostClassifier ` is limited to  "reg:squarederror", "reg:tweedie", "reg:linear", "reg:logistic", "binary:logistic", "binary:logitraw". And these values are even forced with validation code:
```
        if kkey == "objective":
            if not isinstance(parameters["objective"], str):
                raise TypeError("'objective' must be of type str")
            if parameters["objective"] not in [
                "reg:squarederror",
                "reg:tweedie",
                "reg:linear",
                "reg:logistic",
                "binary:logistic",
                "binary:logitraw",
            ]:
                raise ValueError(
                    """'objective' must either be 'reg:squarederror', """
                    """'reg:tweedie', 'reg:linear', 'reg:logistic', """
                    """'binary:logistic', or 'binary:logitraw'"""
                )
```
This code was clearly added when XGBoost already supported multiclass classification. So why can't I use an objective like "multi:softmax"? Or maybe there is some workaround for the multiclass classification?

**Comment by liuzicheng1987** [comment]
Hi @paulbir,

the issue is not so much XGBoost itself, but the feature learning algorithms. It can be very tricky to build features for very high-dimensional targets.

You can use the getml.data.make_target_columns(…), as exemplified in the CORA notebook:

https://nbviewer.org/github/getml/getml-demo/blob/master/cora.ipynb

**Comment by paulbir** [comment]
Hi @liuzicheng1987 , thanks for your reply. 

I have target with 3 classes. This is a multiclass problem, so the objective should be multi:softmax or multi:softprob, but only binary targets are allowed.

**Comment by paulbir** [comment]
Hi @srnnkls . My goal is to create new features using the Relboost method. In all the notebook examples I can see that when creating the `pipeline` like here:
```
pipe = getml.pipeline.Pipeline(
    data_model=time_series.data_model,
    tags=["memory=15", "logistic regression"],
    feature_learners=[feature_learner],
    predictors=[predictor],
)
```
And the docs do not explicitly state if the predictor is necessary for feature engineering itself. So I usually set it with `getml.predictors.XGBoostClassifier`. I assumed that it is necessary in some intermediate steps internally.

But what I thought now is do I really need to set the `predictors` parameter for feature engineering only?

**Comment by liuzicheng1987** [comment]
@paulbir , no, if all you are interested in are the features, you don't really need predictors.

**Comment by paulbir** [comment]
@liuzicheng1987 thanks. So I have no issues with predictors anymore then.

**Comment by Jogala** [comment]
Let is define the number of class labels as L.

@paulbir just to clarify:

You can do multiclass classification using getml and `getml.predictors.XGBoostClassifier(objective="binary:logistic")`. What getml does, it builds a set of features for each class label and trains a separate predictor for each of them. It performs a 1 vs all approach for each class label. If you call `v=ipe.predict(container.test)` on such a pipeline, then `v_ij` is the probability that the i-th row belongs to the label j and not any other label. This procedure requires declaring multiple targets as @liuzicheng1987 mentioned in his answer using  the make_target_columns function.

You can also follow @srnnkls approach, that you use the getml pipeline for constructing the features and then use `pipe.transform(container.test)` to generate the flat population table containing the combined L features sets. Following that, you can train a XGBoost predictor using `"objective": "multi:softmax"``. Note however that you can not use a new split! You have to use the test/train partition that you used for generation the features in first place. Otherwise you will get a leak from the training data into the test data.

Example can be found here: https://github.com/Jogala/cora under scripts/ml_all.py
